File: core/src/io/github/jerukan/util/heightmaps/DiamondSquare.java
package io.github.jerukan.util.heightmaps;

import com.badlogic.gdx.math.MathUtils;
import io.github.jerukan.game.board.Tile;
import io.github.jerukan.util.Constants;
import io.github.jerukan.util.Position;
import io.github.jerukan.util.Util;

import java.util.ArrayList;

public class DiamondSquare {

    public static double[][] heights = new double[Constants.BOARD_WIDTH][Constants.BOARD_HEIGHT];
    private static double offsetFactorConstant = 0.7;
    private static double offsetFactor = 0.5;

    private static void initCorners() {
        heights[0][0] = Math.random();
        heights[Constants.BOARD_WIDTH - 1][0] = Math.random();
        heights[0][Constants.BOARD_HEIGHT - 1] = Math.random();
        heights[Constants.BOARD_WIDTH - 1][Constants.BOARD_HEIGHT - 1] = Math.random();
    }

    private static void reset() {
        offsetFactor = 0.3;
        for(int x = 0; x < Constants.BOARD_WIDTH; x++) {
            for (int y = 0; y < Constants.BOARD_HEIGHT; y++) {
                heights[x][y] = 0;
            }
        }
        initCorners();
    }

    private static double squareStep(int xpos, int ypos, int radius) {
        ArrayList<Position> points = new Position(xpos, ypos).getPositionsDistance(radius);
        ArrayList<Double> selectHeights = new ArrayList<>();
        for(Position p : points) {
            if(heights[p.getX()][p.getY()] == 0) {
                return 0; //if any of the heights are 0 return no height
            }
            selectHeights.add(heights[p.getX()][p.getY()]);
        }
        double offset = (2 * Math.random() - 1) * offsetFactor;
        offsetFactor *= offsetFactorConstant;
        return Util.averageList(selectHeights) + offset;
    }

    private static double diamondStep(int xpos, int ypos, int radius) {
        ArrayList<Position> points = new Position(xpos, ypos).getPositionsDiagonal(radius, radius);
        ArrayList<Double> selectHeights = new ArrayList<>();
        for(Position p : points) {
            if(heights[p.getX()][p.getY()] == 0) {
                return 0; //if any of the heights are 0 return no height
            }
            selectHeights.add(heights[p.getX()][p.getY()]);
        }
        double offset = (2 * Math.random() - 1) * offsetFactor;
        offsetFactor *= offsetFactorConstant;
        return Util.averageList(selectHeights) + offset;
    }

    public static void generateHeights() {
        reset();
        for(int radius = (Constants.BOARD_WIDTH - 1) / 2; radius > 0; radius--) {
            for(int x = 0; x < Constants.BOARD_WIDTH; x++) {
                for(int y = 0; y < Constants.BOARD_HEIGHT; y++) {
                    if(heights[x][y] == 0) {
                        heights[x][y] = MathUtils.clamp(diamondStep(x, y, radius), 0 ,1);
                    }
                }
            }
            for(int x = 0; x < Constants.BOARD_WIDTH; x++) {
                for(int y = 0; y < Constants.BOARD_HEIGHT; y++) {
                    if(heights[x][y] == 0) {
                        heights[x][y] = MathUtils.clamp(squareStep(x, y, radius), 0, 1);
                    }
                }
            }
        }
    }
}

File: core/src/io/github/jerukan/util/heightmaps/DiamondSquare.kt
package io.github.jerukan.util.heightmaps

import com.badlogic.gdx.math.MathUtils
import io.github.jerukan.util.Constants
import io.github.jerukan.util.Position
import io.github.jerukan.util.Util

import java.util.ArrayList
import java.util.Random

object DiamondSquare {

    private var diamondsteps = 0
    private var squaresteps = 0

    var heights = Array(Constants.BOARD_WIDTH) { DoubleArray(Constants.BOARD_HEIGHT) }
    private val OFFSET_FACTOR_DEFAULT = 0.5
    private val OFFSET_MULTIPLIER = 0.7
    private var offsetFactor = OFFSET_FACTOR_DEFAULT

    private val DEFAULT_SEED: Long = 1
    private val random = Random(DEFAULT_SEED)

    fun setSeed(seed: Long) {
        random.setSeed(seed)
    }

    private fun initCorners() {
        heights[0][0] = random.nextDouble()
        heights[Constants.BOARD_WIDTH - 1][0] = random.nextDouble()
        heights[0][Constants.BOARD_HEIGHT - 1] = random.nextDouble()
        heights[Constants.BOARD_WIDTH - 1][Constants.BOARD_HEIGHT - 1] = random.nextDouble()
    }

    private fun reset() {
        offsetFactor = OFFSET_FACTOR_DEFAULT
        for (x in 0 until Constants.BOARD_WIDTH) {
            for (y in 0 until Constants.BOARD_HEIGHT) {
                heights[x][y] = 0.0
            }
        }
        initCorners()
    }

    //apparently this is actually a diamond step but whatever
    private fun squareStep(xpos: Int, ypos: Int, radius: Int): Double {
        val points = Position(xpos, ypos).getPositionsDistance(radius)
        val selectHeights = ArrayList<Double>()
        for (p in points) {
            if (heights[p.x][p.y] == 0.0) {
                return 0.0 //if any of the heights are 0 return no height
            }
            selectHeights.add(heights[p.x][p.y])
        }
        val offset = (2 * random.nextDouble() - 1) * offsetFactor
        squaresteps++
        return Util.averageList(selectHeights) + offset
    }

    //this is supposed to be the square step (?)
    private fun diamondStep(xpos: Int, ypos: Int, radius: Int): Double {
        val points = Position(xpos, ypos).getPositionsDiagonal(radius, radius)
        val selectHeights = ArrayList<Double>()
        for (p in points) {
            if (heights[p.x][p.y] == 0.0) {
                return 0.0 //if any of the heights are 0 return no height
            }
            selectHeights.add(heights[p.x][p.y])
        }
        val offset = (2 * random.nextDouble() - 1) * offsetFactor
        diamondsteps++
        return Util.averageList(selectHeights) + offset
    }

    fun generateHeights() {
        reset()
        for (radius in (Constants.BOARD_WIDTH - 1) / 2 downTo 1) {
            diamondsteps = 0
            squaresteps = 0
            for (x in 0 until Constants.BOARD_WIDTH) {
                for (y in 0 until Constants.BOARD_HEIGHT) {
                    if (heights[x][y] == 0.0) {
                        heights[x][y] = MathUtils.clamp(diamondStep(x, y, radius), 0.0, 1.0)
                    }
                }
            }
            for (x in 0 until Constants.BOARD_WIDTH) {
                for (y in 0 until Constants.BOARD_HEIGHT) {
                    if (heights[x][y] == 0.0) {
                        heights[x][y] = MathUtils.clamp(squareStep(x, y, radius), 0.0, 1.0)
                    }
                }
            }
            //            System.out.println("diamond steps: " + diamondsteps);
            //            System.out.println("square steps: " + squaresteps);
            offsetFactor *= OFFSET_MULTIPLIER
        }
    }
}
